Integrations/Snowflake/Snowflake.py

```diff
@@ -0,0 +1,392 @@
+        to which it belongs
+
+    returns:
+        Incident Object
+    """
+    incident = {}
+    occurred = row.get(DATETIME_COLUMN)
+    timestamp = None
+    if occurred:
+        if isinstance(occurred, dttime) or isinstance(occurred, timedelta):
+            err_msg = 'The datetime field specified in the integration parameters must '
+            err_msg += 'contain values of type "datetime" or "date".'
+            raise FetchError(err_msg)
+        timestamp = occurred.timestamp() * 1000
+    else:
+        err_msg = 'Nothing found when trying to fetch the datetime field specified in'
+        err_msg += ' the integration parameters. Please check that the name was correct.'
+        err_msg += ' If the field name was correct, verify that the returned value for'
+        err_msg += ' the specified field is not NULL for ALL of the rows to be fetched.'
+        raise FetchError(err_msg)
+    # Incident Title
+    if INCIDENT_NAME_COLUMN:
+        name = row.get(INCIDENT_NAME_COLUMN)
+    else:
+        name = 'Snowflake Incident -- ' + convert_datetime_to_string(occurred) + '-' + datetime.now().timestamp()
+    incident['name'] = name
+    # Incident occurrence time as timestamp - the datetime field specified in the integration parameters
+    incident['timestamp'] = timestamp
+    # The raw response for the row (reformatted to be json serializable) returned by the db query
+    reformatted_row = format_to_json_serializable(column_descriptions, row)
+    incident['rawJSON'] = json.dumps(reformatted_row)
+    return incident
+
+
+'''MAIN FUNCTIONS / API CALLS'''
+
+
+def test_module():
+    """
+    Test the validity of the integration instance parameters by trying to create a connection
+
+    returns:
+        An 'ok' message if valid, otherwise an error message
+    """
+    conn = get_connection({})
+    demisto.results('ok')
+    conn.close()
+
+
+def fetch_incidents():
+    """
+    Fetch events from this integration and return them as Demisto incidents
+
+    returns:
+        Demisto incidents
+    """
+    # demisto.getLastRun() will returns an obj with the previous run in it.
+    last_run = demisto.getLastRun()
+    # Get the last fetch time, if exists
+    last_fetch = last_run.get('last_fetched_data_timestamp')
+    first_fetch = False
+
+    # Handle first time fetch, fetch incidents retroactively
+    if not last_fetch:
+        first_fetch = True
+        last_fetch, _ = parse_date_range(FETCH_TIME, to_timestamp=True)
+    args = {'rows': MAX_ROWS, 'query': FETCH_QUERY}
+
+    column_descriptions, data = snowflake_query(args)
+    # convert the data/events to demisto incidents
+    incidents = []
+    for row in data:
+        incident = row_to_incident(column_descriptions, row)
+        incident_timestamp = incident.get('timestamp')
+
+        # Update last run and add incident if the incident is newer than last fetch
+        if incident_timestamp and incident_timestamp >= last_fetch:
+            if first_fetch:
```

Comment:
can't we query by the timestamp and avoid all that ?

Reply:
Not if we allow for identical timestamps for different alerts as I described that I was doing in the above comment